# mmistat

[minimap2](https://github.com/lh3/minimap2) index file (.mmi) statistics

[![build](https://github.com/bio-cr/mmistat/actions/workflows/build.yml/badge.svg)](https://github.com/bio-cr/mmistat/actions/workflows/build.yml)
[![test](https://github.com/bio-cr/mmistat/actions/workflows/test.yml/badge.svg)](https://github.com/bio-cr/mmistat/actions/workflows/test.yml)

## Installation

Download the latest release from [GitHub Releases](https://github.com/bio-cr/mmistat/releases).

Note: Binaries for MacOS are not statically linked. To use on macOS, please build from source.

## Usage

Create a minimap2 index file with the `-d` option:

```sh
# For Oxford Nanopore reads
minimap2 -x map-ont -d MT-human-ont.mmi test/MT-human.fa
# For PacBio reads
minimap2 -x map-pb  -d MT-human-pb.mmi  test/MT-human.fa
```

Then run `mmistat` on the map-ont index file:

```sh
mmistat MT-human-ont.mmi
```

```console
Magic number (raw): "MMI\u0002"
MMI Header:
  k-mer size: 15
  Seed width: 10
  Bucket bits: 14
  Number of sequences: 1
  Flags: 0x0

Sequences:
  Idx   Name    Length
  [1]   MT_human        16569
```

Run `mmistat` on the map-pb index file:

```sh
mmistat MT-human-pb.mmi
```

```console
Magic number (raw): "MMI\u0002"
MMI Header:
  k-mer size: 19
  Seed width: 10
  Bucket bits: 14
  Number of sequences: 1
  Flags: 0x1

Sequences:
  Idx   Name    Length
  [1]   MT_human        16569
```

You can see that the k-mer size differs between the two index files.

The output also shows the names and lengths of the chromosomes.

This allows you to verify that the index file is as expected.

## Build

### Crystal version

The Crystal version is faster because it only parses the header.

It is expected to output the same results as the C version.

```sh
crystal build mmistat.cr
```

### C version (for testing)

The C version is reliable because it uses minimap2 as a library.

```sh
make
```

## License

MIT

The project includes code generated by AI.
